### PR TITLE
DEPR: fillna downcasting from object dtype

### LINKED ISSUE #11537: BUG/API: Clarify the behaviour of fillna downcasting
From #11497. Found inconsistencyr regarding `fillna` downcasting.

I understand `fillna(downcast=None)` should downcast values appropriately. It doesn't work on `float` dtype.

```
# NG
pd.Series([1, 2, np.nan, 4]).fillna(3)
#0    1
#1    2
#2    3
#3    4
# dtype: float64

# OK
pd.Series([1, 2, np.nan, 4], dtype=object).fillna(3)
#0    1
#1    2
#2    3
#3    4
# dtype: int64
```

Based on the internal logic, `downcast` can accept dtype. It works on `float`, but not on `object`.
- https://github.com/pydata/pandas/blob/master/pandas/core/internals.py#L366

```
# OK
pd.Series([1, 2, np.nan, 4]).fillna(3, downcast='int')
#0    1
#1    2
#2    3
#3    4
# dtype: int64

# NG
pd.Series([1, 2, np.nan, 4], dtype=object).fillna(3, downcast='int')
#0    1
#1    2
#2    3
#3    4
# dtype: object
```

I understood the expected behavior as below:
- all dtypes should be downcast by default (downcast=None)
- if any dtype is passed via `downcast` kw, downcast to the specified dtype if possible. 
- `downcast=False` will not perform downcast. 
- [ ] Add `Index.fillna` to API (follow-up of #11343)

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #11537 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by mroeschke** [comment]
```
2023-07-26T01:20:55.4868437Z RuntimeError: Non Expected warning in `/home/runner/work/pandas/pandas/doc/source/user_guide/missing_data.rst` line 709
```

File: doc/source/user_guide/missing_data.rst [hunk 1]
```diff
@@ -689,6 +689,7 @@ contains NAs, an exception will be generated:
 However, these can be filled in using :meth:`~DataFrame.fillna` and it will work fine:
 
 .. ipython:: python
+   :okwarning:
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply doc/source/user_guide/missing_data.rst:692]: wouldn't it be weird to have documented behavior that you cant actually get yet?

[the remark this answers is not in the document]
Comment (mroeschke) [reply doc/source/user_guide/missing_data.rst:692]: Or rather, can this be changed to show the work-around? (Just trying to get ahead when we will have to change this example eventually)

[the remark this answers is not in the document]
Comment (jbrockmendel) [reply doc/source/user_guide/missing_data.rst:692]: Fair enough, I'll take a look in a bit. (Pretty deep in the weeds in the extension tests ATM)

File: doc/source/whatsnew/v2.1.0.rst [hunk 2]
```diff
@@ -592,6 +592,7 @@ Other Deprecations
 - Deprecated the behavior of :meth:`Index.argmax`, :meth:`Index.argmin`, :meth:`Series.argmax`, :meth:`Series.argmin` with either all-NAs and ``skipna=True`` or any-NAs and ``skipna=False`` returning -1; in a future version this will raise ``ValueError`` (:issue:`33941`, :issue:`33942`)
 - Deprecated allowing non-keyword arguments in :meth:`DataFrame.to_sql` except ``name`` and ``con`` (:issue:`54229`)
 - Deprecated silently ignoring ``fill_value`` when passing both ``freq`` and ``fill_value`` to :meth:`DataFrame.shift`, :meth:`Series.shift` and :meth:`.DataFrameGroupBy.shift`; in a future version this will raise ``ValueError`` (:issue:`53832`)
+-
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.0.rst:595]:
```suggestion
```

**Comment by jbrockmendel** [comment]
updated+green

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Pull request merged**